hal/src/samd21/usb/bus.rs

```diff
@@ -152,12 +152,13 @@ impl AllEndpoints {
 
 // FIXME: replace with more general heap?
 const BUFFER_SIZE: usize = 2048;
-fn buffer() -> &'static mut [u8; BUFFER_SIZE] {
-    singleton!(: [u8; BUFFER_SIZE] = unsafe{MaybeUninit::uninit().assume_init()}).unwrap()
+fn buffer() -> &'static mut [MaybeUninit<u8>; BUFFER_SIZE] {
+    singleton!(: [MaybeUninit<u8>; BUFFER_SIZE] = unsafe{ MaybeUninit::uninit().assume_init() })
+        .unwrap()
 }
 
 struct BufferAllocator {
-    buffers: &'static mut [u8; BUFFER_SIZE],
+    buffers: &'static mut [MaybeUninit<u8>; BUFFER_SIZE],
```

Comment:
What's the point of `MaybeUninit<u8>`? `u8` is always valid. If anything, I would have expected `MaybeUninit<[u8; BUFFER_SIZE]>`. This would prevent the buffer to be zero-initialized, but I think this would be UB, because the buffer is not actually initialized, before `assume_init()` is called.

Reply:
Apparently what we had was UB. This is the suggested fix as far as I can tell. I could try changing it to `MaybeUninit<[u8; BUFFER_SIZE]>`. I didn't try that.

Below is the clippy lint that motivated the changes to this buffer code:
```rust
error: this call for this type may be undefined behavior
1305
   --> /home/runner/work/atsamd/atsamd/hal/src/samd21/usb/bus.rs:156:45
1306
    |
1307
156 |     singleton!(: [u8; BUFFER_SIZE] = unsafe{MaybeUninit::uninit().assume_init()}).unwrap()
1308
    |                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
1309
    |
1310
    = note: `#[deny(clippy::uninit_assumed_init)]` on by default
1311
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#uninit_assumed_init
```

Reply:
But the new thing is probably just as much UB. To be safe I would write `singleton!(: [u8; BUFFER_SIZE] = [0u8; BUFFER_SIZE]).unwrap()` since the original author did not add a comment to explain why this may be safe.
The downside would only be a few microseconds at reset to zero-initialize this.

Reply:
I don't know enough about `MaybeUninit` to know what is best to do here, so deferring to your suggestion!  Will add a TODO item.

Reply:
This buffer is a backing buffer for usb packets and endpoint descriptors. Data is never read from it that is not previously written with known values. The reads don't extend beyond the known values. Hence the `assume_init`

Reply:
Specifically, the buffer is used to store the inbound/outbound USB frames. On the inbound side, data is only read up to the size that the USB peripheral has written: https://github.com/atsamd-rs/atsamd/blob/0fc72b3ca4c363e158350a88ff7b85f63327abde/hal/src/samd21/usb/bus.rs#L447

On the outbound side, we write the data so its implicitly initialized: https://github.com/atsamd-rs/atsamd/blob/0fc72b3ca4c363e158350a88ff7b85f63327abde/hal/src/samd21/usb/bus.rs#L338

Reply:
Looking at the documentation for [`assume_init`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.assume_init), it looks like we must call it only after it has been initialized. This is not the case here. I think I agree with @Sympatron , the amount of time it takes to initialize the buffer once on boot is trivial and just avoid `MaybeUninit` and some `unsafe` code all together. Any objections?

Reply:
Good catch on the semantics for `assume_init` - SGTM on initializing the buffer.

Reply:
updated with this change